[JSC] Add fast path for `Array.from(arguments)`

sosukesuzuki opened the pull request:
#### 12bbbd4b4f3c62309dd8bbd5acd81e8e46b96aab
<pre>
[JSC] Add fast path for `Array.from(arguments)`
<a href="https://bugs.webkit.org/show_bug.cgi?id=302236">https://bugs.webkit.org/show_bug.cgi?id=302236</a>

Reviewed by Yusuke Suzuki.

This patch changes to add fast path for `Array.from(arguments)`.

note: This patch only supports `DirectArguments` and `ScopedArguments`.
I&apos;ll add support for `ClonedArguments` in next patch.

                                        TipOfTree                  Patched

array-from-scoped-arguments           9.1958+-0.1530     ^      3.6832+-0.2937        ^ definitely 2.4967x faster
array-from-direct-arguments           8.2003+-0.1391     ^      2.8755+-0.0549        ^ definitely 2.8518x faster

Test: JSTests/microbenchmarks/array-from-direct-arguments.js

* JSTests/microbenchmarks/array-from-direct-arguments.js: Added.
(test):
* JSTests/microbenchmarks/array-from-scoped-arguments.js: Added.
(test):
* JSTests/stress/array-from-direct-arguments.js: Added.
(shouldBe):
(test):
(noInline):
* JSTests/stress/array-from-scoped-arguments.js: Added.
(shouldBe):
(test):
(noInline):
* Source/JavaScriptCore/builtins/ArrayConstructor.js:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/ArrayConstructor.cpp:
(JSC::tryCreateArrayFromArguments):
(JSC::tryCreateArrayFromScopedArguments):
(JSC::tryCreateArrayFromDirectArguments):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ArrayConstructor.h:
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
* Source/JavaScriptCore/runtime/ArrayPrototype.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):

Canonical link: <a href="https://commits.webkit.org/302891@main">https://commits.webkit.org/302891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ad81ad721876a7c22b1bfad012ee7b8502126b0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130222 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2493 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41176 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137640 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81793 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2574 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2383 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99204 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67062 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133169 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1897 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116623 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79895 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1819 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34751 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80896 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122225 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110363 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35258 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140117 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128674 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2283 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2123 "Found 1 new test failure: imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/selectors/cue/white-space_pre.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107727 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2327 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112966 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107615 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27458 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1805 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31416 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55229 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2353 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65740 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161689 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2170 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40322 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2374 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2279 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/builtins/ArrayConstructor.js:
```diff
@@ -44,6 +44,9 @@ function from(items /*, mapFn, thisArg */)
         var fastResult = @arrayFromFastFillWithUndefined(this, arrayLike);
         if (fastResult)
             return fastResult;
+        fastResult = @arrayFromFastArguments(this, arrayLike);
```
I think we should merge both given that both functions are only called from here.

sosukesuzuki replied:
Removed `arrayFromFastFillWithUndefined` and merged it into the new `arrayFromFastWithoutMapFn` function. Also, since `Array.from` is a constructor function, it is defined in `ArrayConstructor.cpp` and not `ArrayPrototype.cpp`.

sosukesuzuki committed:
[JSC] Add fast path for `Array.from(arguments)`
https://bugs.webkit.org/show_bug.cgi?id=302236

Reviewed by Yusuke Suzuki.

This patch changes to add fast path for `Array.from(arguments)`.

note: This patch only supports `DirectArguments` and `ScopedArguments`.
I'll add support for `ClonedArguments` in next patch.

                                        TipOfTree                  Patched

array-from-scoped-arguments           9.1958+-0.1530     ^      3.6832+-0.2937        ^ definitely 2.4967x faster
array-from-direct-arguments           8.2003+-0.1391     ^      2.8755+-0.0549        ^ definitely 2.8518x faster

Test: JSTests/microbenchmarks/array-from-direct-arguments.js

* JSTests/microbenchmarks/array-from-direct-arguments.js: Added.
(test):
* JSTests/microbenchmarks/array-from-scoped-arguments.js: Added.
(test):
* JSTests/stress/array-from-direct-arguments.js: Added.
(shouldBe):
(test):
(noInline):
* JSTests/stress/array-from-scoped-arguments.js: Added.
(shouldBe):
(test):
(noInline):
* Source/JavaScriptCore/builtins/ArrayConstructor.js:
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/ArrayConstructor.cpp:
(JSC::tryCreateArrayFromArguments):
(JSC::tryCreateArrayFromScopedArguments):
(JSC::tryCreateArrayFromDirectArguments):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ArrayConstructor.h:
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
* Source/JavaScriptCore/runtime/ArrayPrototype.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):

Canonical link: https://commits.webkit.org/302891@main
